### ISSUE TITLE
(Nice to have) Feature request: truly non-blocking `Produce` method

### ISSUE THREAD

**Issue by beanliu** [twmb/franz-go#132]
The current implementation of the [Client.Produce](https://pkg.go.dev/github.com/twmb/franz-go@v1.3.1/pkg/kgo#Client.Produce) will block if the buffer is full. For producers that are not on the critical path of a service, it would be nice to have something like `TryProduce`, which returns immediately if there's no space left in the buffer. So that the caller could drop the record and continue.

A bit more context about the use case: the producer runs in a resource-bounded environment and is used to report events that are not critical (but nice to have). The expectation is that it will try its best to produce, and the caller will keep track of the number of successes/failures.

The current solution uses a buffered channel as a middle layer between the caller and the producer, `select-with-default` as the non-blocking mechanism. It works, but still, it would be great to have a more straightforward solution. `BufferedProduceRecords()` looks good, but there is no way to guarantee the non-blocking property under heavy concurrency.

**Comment by twmb** [comment]
What do you think about having this be a configuration option, rather than a new method on the client?
```go
// NonBlockingProduce configures an auto-flushing client to immediately fail
// with ErrMaxBuffered if the maximum amount of records are buffered,
// overriding the default behavior to block and wait for some records to flush.
//
// This option is implied if using manual flushing.
func NonBlockingProduce() ProducerOpt
```
?

Would there be a reason for TryProduce (non-blocking) and Produce (blocking) in the same client? I'm undecided on whether a third way of producing should exist, because it also implies a fourth synchronous non-blocking produce should exist as well.

**Comment by twmb** [comment]
I'm leaning towards `TryProduce` and `TryProduceSync`. These will be added to `GroupTransactSession` as well, resulting in 4 new methods across two types, but I think the burden of learning is small.

**Comment by beanliu** [comment]
> What do you think about having this be a configuration option, rather than a new method on the client?

Whichever works for me and makes not much difference in my case, so I don't have a strong opinion to make that call from a user's perspective. 

And thanks for all the quick replies and related changes.

**Comment by twmb** [comment]
This will be in 1.4 -- only TryProduce, not TryProduceSync -- it makes less sense to have a potentially failing synchronous method.

**Comment by twmb** [comment]
v1.4 tracked in #135

**Comment by twmb** [comment]
Sorry about the delay -- v1.4 _should be_ this week. I think there's _one_ more thing to add to it. I'll reopen this for now so that it can be properly closed when v1.4 is released (now that I have a v1.4 tracking issue).

**Comment by twmb** [comment]
This has been tagged, v1.4.0.
